cs/src/core/protocols/CompactBinary.cs

```diff
@@ -171,7 +195,16 @@ public void WriteVersion()
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
 #endif
         public void WriteStructBegin(Metadata metadata)
-        {}
+        {
+            if (version == 2)
+            {
+                uint length = lengths[0];
+                lengths.RemoveAt(0);
```

Comment:
`RemoveAt(0)` is going to slide all the lengths down taking O(n) operations, for a total complexity of O(n^2). Did you look at using a different data structure or an index?